### PR TITLE
test: Add Mock server test for cpp

### PR DESCRIPTION
**Pull request by ddiakiteaneo**

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -313,7 +313,7 @@ jobs:
           name: code-coverage-report-xml
           path: packages/python/coverage.xml
 
-  build-cpp-packages:
+  build-test-cpp:
```
Comment (lemaitre-aneo) [inline .github/workflows/ci.yml:316]: The package building and the test should be two separate jobs.

File: packages/cpp/ArmoniK.Api.Tests/source/ChannelOptionsTest.cpp [hunk 2]
```diff
@@ -32,45 +38,53 @@ size_t num_create_session_submitter = 0;
 class Options : public MockFixture {};
 
 TEST_F(Options, no_options) {
+  Logger logger{armonik::api::common::logger::writer_console(), armonik::api::common::logger::formatter_plain(true)};
   armonik::api::common::utils::Configuration configuration;
   configuration.add_json_configuration("appsettings.json").add_env_configuration();
 
   std::string server_address = configuration.get("Grpc__EndPoint");
-  auto channel = ::grpc::CreateChannel(server_address, grpc::InsecureChannelCredentials());
+  // auto args = armonik::api::common::utils::getChannelArguments(configuration);
```
Comment (ngruelaneo) [inline packages/cpp/ArmoniK.Api.Tests/source/ChannelOptionsTest.cpp:46]: pourquoi garder ces commentaires?

File: packages/cpp/ArmoniK.Api.Tests/source/ChannelOptionsTest.cpp [hunk 3]
```diff
@@ -32,45 +38,53 @@ size_t num_create_session_submitter = 0;
...
 
 TEST_F(Options, default_options) {
+  Logger logger{armonik::api::common::logger::writer_console(), armonik::api::common::logger::formatter_plain(true)};
   armonik::api::common::utils::Configuration configuration;
   configuration.add_json_configuration("appsettings.json").add_env_configuration();
 
   std::string server_address = configuration.get("Grpc__EndPoint");
   auto args = armonik::api::common::utils::getChannelArguments(configuration);
-  auto channel = ::grpc::CreateCustomChannel(server_address, grpc::InsecureChannelCredentials(), args);
+  // auto channel = ::grpc::CreateCustomChannel(server_address, grpc::InsecureChannelCredentials(), args);
```
Comment (ngruelaneo) [inline packages/cpp/ArmoniK.Api.Tests/source/ChannelOptionsTest.cpp:64]: idem

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -131,3 +131,78 @@ jobs:
...
           reporter: dotnet-trx
+
+  cpp:
+    strategy:
+      fail-fast: false
+      matrix:
+        endpoint:
+          - {grpc_port: 5000, http_port: 4999, ssl_validation: '', client_cert: ''}
+          - {grpc_port: 5001, http_port: 5001, ssl_validation: enable, client_cert: ''}
+          - {grpc_port: 5003, http_port: 5003, ssl_validation: enable, client_cert: 'true'}
```
Comment (lemaitre-aneo) [inline .github/workflows/test.yml:142]:
```suggestion
          - mock_env:
              Grpc__Port: 5000
              Http_Port: 4999
            test_env:
              Grpc__Endpoint: http://localhost:5000
          - mock_env:
              Grpc__Port: 5001
              Http_Port: 5001
              Http__Cert: $CertFolder/server1.pem
              Http__Key=$CertFolder/server1.key
            test_env:
              Grpc__Endpoint: https://localhost:5001
              Grpc__SSLValidation: enable
              Grpc__CaCert: /app/source/certs/server1-ca.pem
```

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -131,3 +131,78 @@ jobs:
...
+      - name: Build Mock server
+        working-directory: packages/csharp/ArmoniK.Api.Mock
+        shell: bash
+        run: |
+          dotnet publish -o ../out
+          [ -e ../out/ArmoniK.Api.Mock.exe ] || ln -s ArmoniK.Api.Mock ../out/ArmoniK.Api.Mock.exe
+
+      - name: Run Mock servers
+        working-directory: packages/csharp/
+        shell: bash
```
Comment (lemaitre-aneo) [inline .github/workflows/test.yml:182]:
```suggestion
        shell: bash
        env: $(( matrix.endpoint.mock_env ))
```

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -131,3 +131,78 @@ jobs:
...
+          docker run --rm -t --network host \
+            -e Grpc__EndPoint=$( [ ${{ matrix.endpoint.grpc_port }} -eq 5000 ] && echo "http" || echo "https" )://localhost:${{ matrix.endpoint.grpc_port }} \
+            -e Http__EndPoint=$( [ ${{ matrix.endpoint.grpc_port }} -eq 5000 ] && echo "http" || echo "https" )://localhost:${{ matrix.endpoint.http_port }} \
+            $([[ "${{ matrix.endpoint.ssl_validation }}" == "enable" ]] && echo "-e Grpc__SSLValidation=enable -e Grpc__CaCert=/app/source/certs/server1-ca.pem") \
+            $([[ "${{ matrix.endpoint.grpc_port }}" == "5003" ]] && echo "-e Grpc__mTLS=true -e Grpc__CaCert=/app/source/certs/server1.pem -e Grpc__ClientCert=/app/source/certs/client.pem -e Grpc__ClientKey=/app/source/certs/client.key") \
+            "armonik-api-cpp:0.1.0"
+
+      - name: Stop Mock Server
+        run: |
+          ps -p $server_pid > /dev/null && kill $server_pid
```
Comment (lemaitre-aneo) [inline .github/workflows/test.yml:208]:
```suggestion
          kill $server_pid || true
```

File: .github/workflows/test.yml [hunk 7]
```diff
@@ -131,3 +131,78 @@ jobs:
...
+        run: |
+          ./run_test.sh
+
+      - name: Run Test
+        run: |
+          docker run --rm -t --network host \
+            -e Grpc__EndPoint=$( [ ${{ matrix.endpoint.grpc_port }} -eq 5000 ] && echo "http" || echo "https" )://localhost:${{ matrix.endpoint.grpc_port }} \
+            -e Http__EndPoint=$( [ ${{ matrix.endpoint.grpc_port }} -eq 5000 ] && echo "http" || echo "https" )://localhost:${{ matrix.endpoint.http_port }} \
+            $([[ "${{ matrix.endpoint.ssl_validation }}" == "enable" ]] && echo "-e Grpc__SSLValidation=enable -e Grpc__CaCert=/app/source/certs/server1-ca.pem") \
+            $([[ "${{ matrix.endpoint.grpc_port }}" == "5003" ]] && echo "-e Grpc__mTLS=true -e Grpc__CaCert=/app/source/certs/server1.pem -e Grpc__ClientCert=/app/source/certs/client.pem -e Grpc__ClientKey=/app/source/certs/client.key") \
```
Comment (lemaitre-aneo) [inline .github/workflows/test.yml:203]:
```suggestion
            -e "Grpc__Endpoint=$Grpc__Endpoint" -e "Http__EndPoint=${Http__EndPoint:-$Grpc__Endpoint}" \
```

File: packages/cpp/tools/run_test.sh
Comment (lemaitre-aneo) [inline packages/cpp/tools/run_test.sh:1]: Add shebang

```sh
#! /bin/sh
```

File: .github/workflows/test.yml [hunk 8]
```diff
@@ -131,3 +131,145 @@ jobs:
...
+      - name: Run Test
+        env: ${{ matrix.endpoint.test_env }}
+        run: |
+          env
+          docker run --rm -t --network host \
+            -e Grpc__EndPoint=$Grpc__Endpoint \
+            -e Http__EndPoint=${Http__Endpoint:-$Grpc__Endpoint} \
+            -e Grpc__SSLValidation=$Grpc__SSLValidation -e Grpc__CaCert=$Grpc__CaCert \
+            -e Grpc__mTLS=$Grpc__mTLS -e Grpc__ClientCert=$Grpc__ClientCert -e Grpc__ClientKey=$Grpc__ClientKey \
+            "armonik-api-cpp:0.1.0"
```
Comment (lemaitre-aneo) [inline .github/workflows/test.yml:271]:
```suggestion
          docker run --rm -t --network host \
            -e Grpc__EndPoint="$Grpc__Endpoint" \
            -e Http__EndPoint="${Http__Endpoint:-$Grpc__Endpoint}" \
            -e Grpc__SSLValidation="$Grpc__SSLValidation" \
            -e Grpc__CaCert="$Grpc__CaCert" \
            -e Grpc__mTLS="$Grpc__mTLS" \
            -e Grpc__ClientCert="$Grpc__ClientCert" \ 
            -e Grpc__ClientKey="$Grpc__ClientKey" \
            "armonik-api-cpp:0.1.0"
```

File: .github/workflows/test.yml [hunk 9]
```diff
@@ -131,3 +131,145 @@ jobs:
...
+          [ -e ../out/ArmoniK.Api.Mock.exe ] || ln -s ArmoniK.Api.Mock ../out/ArmoniK.Api.Mock.exe
+
+      - name: Run Mock servers
+        working-directory: packages/csharp/
+        shell: bash
+        env: ${{ matrix.endpoint.mock_env }}
+        run: |
+          env
+          export CertFolder="$PWD/../cpp/certs"
+          ./out/ArmoniK.Api.Mock.exe grpc:port=$Grpc__Port http:port=$Http_Port &
```
Comment (lemaitre-aneo) [inline .github/workflows/test.yml:253]:
```suggestion
          ./out/ArmoniK.Api.Mock.exe &
```

**Review by lemaitre-aneo — DISMISSED** [review]

**Review by lemaitre-aneo — APPROVED** [review]

**Pull request merged**